### PR TITLE
MoE fp8 backward: in-kernel drhs operand quantization + configurable gradient dtype

### PR DESCRIPTION
**Pull request by ultrons**
Stacked on #4895 (token quantized all-gather) and #4969 (ring collectives + x_sorted save); the base branch is #4895's head, so the diff shows this change plus #4969's commit — review the top commit.

## What

Two default-off knobs for performance improvement of fp8 MoE training at dropless (worst-case) ragged buffer sizes:

- **`moe_bwd_inkernel_quant`** — quantize the drhs tgmm's operands *inside* the kernel (per-gm-tile-per-channel e4m3, implemented in `tgmm_v2`'s `quantize_operands` path) instead of with dense XLA-level quantize ops. The dense quantize/amax ops process every row of the ragged buffer at its **static** size — at `ragged_buffer_factor<=0` (dropless worst case) that is `tokens * top_k` rows regardless of how many are valid — while the in-kernel path touches only the tiles covered by `group_sizes`. It subsumes three buffer-sized ops: the per-row lhs re-quantize, the `drhs_dout *= lhs.scale` multiply, and the per-N cotangent quantize. A second effect: because no reduction ever reads past the valid rows, the amax can never ingest uninitialized buffer tail rows (with the dense path, a stale-row `amax=Inf` propagates `Inf*0=NaN` into the weight gradient; we root-caused a training NaN to exactly this).
- **`bwd_quantization_dtype`** — the gradient fp8 dtype for the `fp8_full` qwix recipe: `e5m2` (default, existing behavior) or `e4m3`.

Requires `use_tokamax_gmm=true use_gmm_v2=true` and an fp8 qwix `bwd_qtype`; falls back to the XLA-level quantize otherwise. Flags off reproduce the existing behavior.

## Repro

DeepSeek-V3 671B on v7x, the #4895 recipe at the dropless configuration this lever targets:

```
python3 -m maxtext.trainers.pre_train.train maxtext/configs/base.yml \
  model_name=deepseek3-671b per_device_batch_size=1.0 max_target_length=4096 \
  ici_fsdp_parallelism=64 ici_expert_parallelism=8 \
  attention=flash sparse_matmul=true megablox=true use_tokamax_gmm=true use_gmm_v2=true \
  use_tokamax_splash=true use_ring_of_experts=true use_ragged_sort=true \
  use_custom_sort_vjp=true ragged_buffer_factor=-1 num_moe_token_chunks=2 \
  remat_policy=custom decoder_layer_input=offload context=device out_proj=device \
  quantization=fp8_full use_qwix_quantization=true \
  weight_quantization_calibration_method=fixed,-224,224 \
  act_quantization_calibration_method=fixed,-224,224 \
  bwd_quantization_calibration_method=absmax \
  moe_quantize_token_all_gather=true \
  moe_bwd_inkernel_quant=true bwd_quantization_dtype=e4m3 \
  dataset_type=synthetic use_random_routing=true steps=20
```

- The win scales with the gap between the buffer's static size and the balanced size; at small `ragged_buffer_factor` (e.g. 2.0) the dense ops are cheap and the flag is roughly neutral. Use it for dropless (`ragged_buffer_factor=-1`) configurations.
- The in-kernel scales are finer-grained than the XLA path's, so losses match the flag-off run closely but not bit-exactly.

## Validation

- AOT compile (virtual `tpu7x` topology) green: flag-on at the dropless `per_device_batch_size=1.0 / ragged_buffer_factor=-1` config above, and flag-off on the unmodified #4895 recipe.
- 512-chip DeepSeek-V3 runs with this path train cleanly at `ragged_buffer_factor=-1` with no buffer sanitization and no non-finite losses; a 500-step real-data curve on this backward matches the dense-quantize reference at every eval point.

### CODE REVIEW

File: src/maxtext/kernels/ring_ag.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+  prev = [None] * len(lanes)
+  for i in range(s - 1):
+    for li, (si, nb, sign, ss, sl) in enumerate(lanes):
+      start = block_start(my - i if sign < 0 else my + i)
+      if prev[li] is not None:
+        prev[li].wait()
+      prev[li] = pltpu.async_remote_copy(
+          o_ref.at[idx(start, ss, sl)], o_ref.at[idx(start, ss, sl)],
+          send_sem.at[si], recv_sem.at[si], device_id=nb, device_id_type=MESH)
```
[the remark this answers is not in the document]
Comment (ultrons) [reply src/maxtext/kernels/ring_ag.py:123]: No race: `pltpu.make_async_remote_copy(...).wait()` waits on **both** semaphores — the send semaphore and the local `recv_sem` (the split `wait_send()` / `wait_recv()` forms exist precisely because plain `wait()` covers both). So at step `i`, `prev[li].wait()` blocks on `recv_sem.at[si]`, which is signaled exactly by the step `i-1` incoming block from the ring neighbor — the block step `i` forwards. The signal/wait accounting balances: one arrival per step per lane, one recv-wait consumed per step, and the trailing `for d in prev: d.wait()` consumes the final arrival. Adding an extra `pl.semaphore_wait(recv_sem.at[si], i)` would double-count the arrivals and deadlock. The kernel is validated bit-exact against `jax.lax.all_gather(..., tiled=True)` on v7x hardware from 8 to 512 chips.

File: src/maxtext/kernels/megablox/ops.py [hunk 2]
```diff
@@ -493,14 +497,36 @@ def _gmm_bwd(
...
+      grad, residual_lhs, residual_rhs, group_sizes, use_gmm_v2, transpose_rhs, quantization_rule,
+      skip_lhs_quant=inkernel_drhs,
   )
 
   # 2. Backward Pass Quantization
   if quantization_rule:
-    dlhs_dout, drhs_dout = _bwd_quantize_gradient(dlhs_dout, drhs_dout, quantization_rule)
+    if inkernel_drhs:
+      # the in-kernel tgmm quantizes drhs_dout ITSELF; keep the raw cotangent for the drhs side.
+      dlhs_dout, _ = _bwd_quantize_gradient(dlhs_dout, drhs_dout, quantization_rule)
```
[the remark this answers is not in the document]
Comment (ultrons) [reply src/maxtext/kernels/megablox/ops.py]: Good catch — the discarded second output still traced the dense drhs quantize (XLA DCE removes it, but relying on DCE for the exact op this flag eliminates is fragile). Applied: the in-kernel branch now quantizes only the dlhs cotangent inline.

File: src/maxtext/configs/types.py [hunk 3]
```diff
@@ -884,6 +884,35 @@ class MoEGeneral(BaseModel):
...
+          "dropless (worst-case) ragged buffer sizes."
+      ),
+  )
+  bwd_quantization_dtype: str = Field(
+      "e5m2",
+      description=(
+          "fp8 dtype for the BACKWARD (gradient) quantization in the fp8_full qwix recipe: 'e5m2' "
+          "(default) or 'e4m3'."
+      ),
+  )
```
[the remark this answers is not in the document]
Comment (ultrons) [reply src/maxtext/configs/types.py]: Applied — `Literal["e5m2", "e4m3"]`.

**Comment by RissyRan** [comment]
Thanks for the change! It seems from @shuningjin's tests, the feature didn't bring up the gain yet, i.e. rbf=2. Potentially more gains from `ragged_buffer_factor=-1` use case. Wondering if you see any in your tests?
